File: src/ciw/model/julia/src/commitments.jl
# SCR execution commitments (execution.commitments), recomputed by the worker
# from the bytes it actually consumed. The host recomputes them again.

const PROGRAM_TAG = "scout.execution.program.v1"
const INPUT_TAG = "scout.execution.input.v1"
const OUTPUT_TAG = "scout.execution.output.v1"
const COMPUTATION_TAG = "scout.execution.computation.v1"
const SPECIFICATION_TAG = "scout.execution.specification.v1"

function scr_canonical(tag::String, fields::Vector{Vector{UInt8}})
    io = IOBuffer()
    tag_bytes = Vector{UInt8}(codeunits(tag))
    write(io, htol(UInt64(length(tag_bytes))))
    write(io, tag_bytes)
    write(io, htol(UInt64(length(fields))))
    for field in fields
        write(io, htol(UInt64(length(field))))
        write(io, field)
    end
    return take!(io)
end

commit_hex(tag::String, fields::Vector{Vector{UInt8}}) = bytes2hex(sha256(scr_canonical(tag, fields)))

specification_identity(program, configuration, input) =
    commit_hex(SPECIFICATION_TAG, Vector{UInt8}[program, configuration, input])
program_identity(program) = commit_hex(PROGRAM_TAG, Vector{UInt8}[program])
input_identity(input) = commit_hex(INPUT_TAG, Vector{UInt8}[input])
output_identity(output) = commit_hex(OUTPUT_TAG, Vector{UInt8}[output])

function computation_identity(program_id::String, input_id::String, output_id::String, exit_code::Integer)
    return commit_hex(COMPUTATION_TAG, Vector{UInt8}[hex2bytes(program_id), hex2bytes(input_id),
                                                     hex2bytes(output_id), reinterpret(UInt8, [htol(UInt32(exit_code))])])
end

File: src/ciw/model/commitments.py
"""SCR execution commitments, byte-for-byte ``execution.commitments``.

CIW recomputes every identity a worker echoes. The encoding is SCR's
``canonical(tag, fields)``: ``len(tag) u64 LE | tag | count u64 LE | (len u64
LE | bytes)*`` hashed with SHA-256. ``tests/test_model_worker.py`` checks these
functions against the pinned SCR checkout when it is supplied, so this is the
same function in the same seam rather than a second identity system.

``SpecificationIdentity`` binds program, configuration and input bytes.
``ComputationIdentity`` binds program, input, output and exit code only; solver
settings are bound through the retained specification link, never implied by
the computation identity alone.
"""
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import struct

PROGRAM_TAG = "scout.execution.program.v1"
INPUT_TAG = "scout.execution.input.v1"
OUTPUT_TAG = "scout.execution.output.v1"
COMPUTATION_TAG = "scout.execution.computation.v1"
SPECIFICATION_TAG = "scout.execution.specification.v1"


def canonical(tag: str, fields) -> bytes:
    tag_bytes = tag.encode("utf-8")
    out = struct.pack("<Q", len(tag_bytes)) + tag_bytes + struct.pack("<Q", len(fields))
    for field in fields:
        out += struct.pack("<Q", len(field)) + field
    return out


def commit_hex(tag: str, fields) -> str:
    return hashlib.sha256(canonical(tag, fields)).hexdigest()


def canonical_u32(value: int) -> bytes:
    return struct.pack("<I", value)


@dataclass(frozen=True)
class Specification:
    """The SCR ``ExecutionSpecification`` triple: a request, not an event."""

    program: bytes
    configuration: bytes
    input_payload: bytes

    def identity(self) -> str:
        return commit_hex(SPECIFICATION_TAG, [self.program, self.configuration, self.input_payload])

    def program_identity(self) -> str:
        return commit_hex(PROGRAM_TAG, [self.program])

    def input_identity(self) -> str:
        return commit_hex(INPUT_TAG, [self.input_payload])


def output_identity(output: bytes) -> str:
    return commit_hex(OUTPUT_TAG, [output])


def computation_identity(program_identity: str, input_identity: str, output_id: str, exit_code: int) -> str:
    return commit_hex(COMPUTATION_TAG, [bytes.fromhex(program_identity), bytes.fromhex(input_identity),
                                        bytes.fromhex(output_id), canonical_u32(exit_code)])
